src/client/ws/manager.rs

```diff
@@ -0,0 +1,278 @@
+type SubscriptionSink = mpsc::Sender<JsonValue>;
+type UnsubscribeMethod = String;
+type RequestId = u64;
+
+/// Manages and monitors JSONRPC v2 method calls and subscriptions.
+pub struct RequestManager {
+	/// List of requests that are waiting for a response from the server.
+	requests: FnvHashMap<RequestId, Kind>,
+	/// Reverse lookup, to find a request ID in constant time by `subscription ID` instead of looking through all requests.
+	subscriptions: HashMap<SubscriptionId, RequestId>,
+}
+
+impl RequestManager {
+	pub fn new() -> Self {
+		Self { requests: FnvHashMap::default(), subscriptions: HashMap::new() }
+	}
+
+	/// Tries to insert a new pending request.
+	pub fn insert_pending_method_call(
+		&mut self,
+		id: u64,
+		send_back: PendingRequestOneshot,
+	) -> Result<(), PendingRequestOneshot> {
+		if let Entry::Vacant(v) = self.requests.entry(id) {
+			v.insert(Kind::PendingMethodCall(send_back));
+			Ok(())
+		} else {
+			Err(send_back)
+		}
+	}
+
+	/// Tries to inserts a new pending subscription.
+	pub fn insert_pending_subscription(
+		&mut self,
+		id: RequestId,
+		send_back: PendingSubscriptionOneshot,
+		unsubscribe_method: UnsubscribeMethod,
+	) -> Result<(), PendingSubscriptionOneshot> {
+		if let Entry::Vacant(v) = self.requests.entry(id) {
+			v.insert(Kind::PendingSubscription((send_back, unsubscribe_method)));
+			Ok(())
+		} else {
+			Err(send_back)
+		}
+	}
+
+	/// Inserts a new active subscription
+	pub fn insert_active_subscription(
+		&mut self,
+		request_id: RequestId,
+		subscription_id: SubscriptionId,
+		send_back: SubscriptionSink,
+		unsubscribe_method: String,
+	) -> Result<(), SubscriptionSink> {
+		if let (Entry::Vacant(request), Entry::Vacant(subscription)) =
+			(self.requests.entry(request_id), self.subscriptions.entry(subscription_id))
+		{
+			request.insert(Kind::Subscription((send_back, unsubscribe_method)));
+			subscription.insert(request_id);
+			Ok(())
+		} else {
+			Err(send_back)
+		}
+	}
+
+	/// Complete a pending subscription.
+	pub fn complete_pending_subscription(
+		&mut self,
+		request_id: RequestId,
+	) -> Option<(PendingSubscriptionOneshot, UnsubscribeMethod)> {
+		match self.requests.entry(request_id) {
+			Entry::Occupied(request) if matches!(request.get(), Kind::PendingSubscription(_)) => {
+				let (_req_id, kind) = request.remove_entry();
+				if let Kind::PendingSubscription(send_back) = kind {
+					Some(send_back)
+				} else {
+					unreachable!();
+				}
+			}
+			_ => None,
+		}
+	}
+
+	/// Complete a pending request.
+	pub fn complete_pending_method_call(&mut self, request_id: RequestId) -> Option<PendingRequestOneshot> {
+		match self.requests.entry(request_id) {
+			Entry::Occupied(request) if matches!(request.get(), Kind::PendingMethodCall(_)) => {
+				let (_req_id, kind) = request.remove_entry();
+				if let Kind::PendingMethodCall(send_back) = kind {
+					Some(send_back)
+				} else {
+					unreachable!();
+				}
```

Comment:
These `unreachable`s are a bit unfortunate; ideally it shouldn't be possible to complete anything that isn't a `Kind::PendingMethodCall` but not sure how to do that. @maciejhirsz thoughts?
 
Is returning `None` better? :/

Reply:
It is not "pretty" but the "state" is checked above `matches!(request.get(), Kind::PendingMethodCall(_))`  and the `unreachable!` should be impossible to happen but not encoded in the type-system.

So, seemed more intuitive to use unreachable because if that happened it's a logic bug but no strong opinions on it.

As we only have the `request ID`, I don't see how to solve it without a runtime check of the state as this does.

Reply:
Yeah I couldn't come up with a better idea either.

Reply:
`unimplemented!()` is probably a better flag

Reply:
If you call `complete_pending_method_call`, but the state is not pending, that sounds like a logic bug to me. I'd rather have an error type here:

```rust
    pub fn complete_pending_method_call(&mut self, request_id: RequestId) -> Result<PendingRequestOneshot, SomeError> {
        match self.requests.entry(request_id) {
            Entry::Occupied(request) => {
                let (_req_id, kind) = request.remove_entry();
                if let Kind::PendingMethodCall(send_back) = kind {
                    Ok(send_back)
                } else {
                    Err(...)
                }
```

Is there a scenario where this method can be called on any other state (or lack of state) under normal operation?

Reply:
Currently,  it works like this:

```rust
 let response = get_reponse();

 match {
   // response on pending subscriptions and pending requests,
   RequestResponse(response) => {
       match manager.get_status(response.id) {
           RequestStatus::PendingRequest => {
              manager.complete_pending_request(request_id);
           }
           RequestStatus::PendingSubscription => {
               manager.complete_pending_subscription(request_id);
           }
           _ => // don't care invalid request ID
      }
   }
   // response on active subscription
   SubscriptionResponse(sub) => {
      if let Some(request_id) = manager.get_request_id_by_subscription(sub.id) { 
          manager.as_active_subscription_mut(request_id).send(sub.response);
      }
   }
 }
```

I would say that we know the state expect for `active subscription` where I assumed that it must be a subscription because we have a separate message for it (the subscription ID should be unique and very unlikely to have duplicates)

However, I did this very "defensively" because I assumed it could be called on non-validated input too but probably fair to assume that the caller has to ensure that it has a valid state.

FWIW, in your example, if I'd call that with a pending subscription instead it would remove that entry in the map because it has the same `request-id` but the "wrong state".

Reply:
True. The `unreachable!`s there are think otherwise I think, I'll have to think about it a bit more.